File: src/dev/npu/NPUDevice.py
# Copyright (c) 2025
# All rights reserved.
#
# A minimal NPU device model interface for gem5.
#
# This model is intentionally simple: it exposes an MMIO region and a
# DMA-capable interface, but does not implement any real computation.

from m5.objects.Device import DmaVirtDevice
from m5.params import *


class NPUDevice(DmaVirtDevice):
    type = "NPUDevice"
    cxx_header = "dev/npu/npu_device.hh"
    cxx_class = "gem5::NPUDevice"

    # Base address and size of the NPU MMIO window.
    pioAddr = Param.Addr("Base physical address of the NPU MMIO region")
    pioSize = Param.Addr(0x1000, "Size of the NPU MMIO region")

    # Simple fixed MMIO latency.
    pioDelay = Param.Latency("10ns", "Latency for NPU MMIO accesses")

File: src/dev/npu/npu_device.cc
/*
 * Minimal NPU device interface for gem5.
 *
 * This model is a thin stub that exposes an MMIO space and a DMA-capable
 * interface via DmaVirtDevice. It does not implement any real NPU
 * functionality and is intended only to validate integration paths.
 */

#include "dev/npu/npu_device.hh"

#include "base/addr_range.hh"
#include "base/logging.hh"
#include "base/trace.hh"
#include "debug/NPUDevice.hh"

namespace gem5
{

NPUDevice::NPUDevice(const Params &p)
  : DmaVirtDevice(p), pioAddr(p.pioAddr), pioSize(p.pioSize),
    pioDelay(p.pioDelay)
{
}

void
NPUDevice::IdentityTranslationGen::translate(Range &range) const
{
    // Identity mapping: treat vaddr as paddr.
    range.paddr = range.vaddr;
    // Use the full remaining size of the region.
    // range.size is already set by the TranslationGen base class.
}

Tick
NPUDevice::read(PacketPtr pkt)
{
    const Addr addr = pkt->getAddr();
    DPRINTF(NPUDevice, "MMIO read addr=%#x size=%u\n", addr, pkt->getSize());

    pkt->makeAtomicResponse();
    return pioDelay;
}

Tick
NPUDevice::write(PacketPtr pkt)
{
    const Addr addr = pkt->getAddr();
    DPRINTF(NPUDevice, "MMIO write addr=%#x size=%u\n", addr, pkt->getSize());

    pkt->makeAtomicResponse();
    return pioDelay;
}

AddrRangeList
NPUDevice::getAddrRanges() const
{
    AddrRangeList ranges;
    if (pioSize != 0) {
        ranges.push_back(RangeSize(pioAddr, pioSize));
    }
    return ranges;
}

TranslationGenPtr
NPUDevice::translate(Addr vaddr, Addr size)
{
    DPRINTF(NPUDevice, "DMA translate vaddr=%#x size=%u\n", vaddr, size);
    return TranslationGenPtr(new IdentityTranslationGen(vaddr, size));
}

} // namespace gem5

